### PR TITLE
[Reg2Mem] Handle CallBr instructions

### LINKED ISSUE #90900: Reg2Mem pass triggered assertion `isa<To>(V al) && "cast<Ty>() argument of incompatible type!"
https://godbolt.org/z/oxW8zbKr6
`opt --passes=reg2mem reduced.ll`
```llvm
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

%struct.__large_struct = type { [100 x i64] }

define fastcc i64 @__se_sys_pselect6() {
  %1 = callbr i64 asm sideeffect "\0A1:\09movq $1,$0\0A .pushsection \22__ex_table\22,\22a\22\0A .balign 4\0A .long (1b) - .\0A .long (${2:l}) - .\0A .long 3 \0A .popsection\0A", "=r,*m,!i,~{dirflag},~{fpsr},~{flags}"(ptr elementtype(%struct.__large_struct) null)
          to label %2 [label %common.ret]

2:                                                ; preds = %0
  %3 = inttoptr i64 %1 to ptr
  %4 = callbr i64 asm sideeffect "\0A1:\09movq $1,$0\0A .pushsection \22__ex_table\22,\22a\22\0A .balign 4\0A .long (1b) - .\0A .long (${2:l}) - .\0A .long 3 \0A .popsection\0A", "=r,*m,!i,~{dirflag},~{fpsr},~{flags}"(ptr elementtype(%struct.__large_struct) null)
          to label %common.ret [label %common.ret]

common.ret:                                       ; preds = %2, %2, %0
  ret i64 0
}
```

### PR DESCRIPTION
**Pull request by XChy**
Fixes #90900

### CODE REVIEW

File: llvm/lib/Transforms/Utils/DemoteRegToStack.cpp [hunk 1]
```diff
@@ -102,9 +102,12 @@ AllocaInst *llvm::DemoteRegToStack(Instruction &I, bool VolatileLoads,
         new StoreInst(&I, Slot, Handler->getFirstInsertionPt());
       return Slot;
     }
+  } else if (InvokeInst *II = dyn_cast<InvokeInst>(&I)) {
+    InsertPt = II->getNormalDest()->getFirstInsertionPt();
+  } else if (CallBrInst *CBI = dyn_cast<CallBrInst>(&I)) {
+    InsertPt = CBI->getDefaultDest()->getFirstInsertionPt();
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/DemoteRegToStack.cpp:108]: I don't think this is right. The result of callbr can be used on all destinations, not just the default destination.

Comment (XChy) [reply llvm/lib/Transforms/Utils/DemoteRegToStack.cpp:108]: Why? As the langref says:
> The output values of a ‘callbr’ instruction are available only to the ‘fallthrough’ block, not to any ‘indirect’ blocks(s).

Comment (nikic) [reply llvm/lib/Transforms/Utils/DemoteRegToStack.cpp:108]: This used to be the case, but the limitation no longer exists. Looks like we forgot to drop this note in LangRef.

Comment (XChy) [reply llvm/lib/Transforms/Utils/DemoteRegToStack.cpp]: Done now. Do we need to drop the note?

Comment (nikic) [reply llvm/lib/Transforms/Utils/DemoteRegToStack.cpp]: Done in https://github.com/llvm/llvm-project/pull/91167.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**